### PR TITLE
fix(ripple): make ripples conform with specs

### PR DESCRIPTION
**Pull request by DevVersion**
This makes the ripple service conform with the specifications and other Material reference implementations.

See https://material.io/guidelines/motion/material-motion.html#material-motion-how-does-material-move

_Reference from the Material specifications_
> Ink ripples confirm user input by immediately expanding outward from the point of touch. The card lifts to indicate an active state.

This means the following:

 * Ripples now trigger on `mousedown`
 * Ripples now persists as long as the element is being hold.
 * No longer adds an unnecessary background ripple.
 * Removes the ugly `scale(0.00001)` for ripple animations

Not only visually the ripple has been changed. The whole renderer has been rewritten and now has a very simple API, that can be easily used by developers.

Also the API has been changed to something more intuitive

```ts
// (pageX, pageY, config)
ripple.createRipple(0, 0, {
  centered: true, 
  speedFactor: 0.1
});
```

_As seen in the example, developers are now able to specify a simple Ripple Config object_

**Live Demo**: https://material2-dashboard.firebaseapp.com/button (temporary)

References #1434

### CODE REVIEW

**Comment by fxck** [comment]
I'm not sure if material1 conforms spec properly, but it's still slightly different(at least on buttons) https://material.angularjs.org/latest/demo/button background color changes slightly immediately after mousedown.

**Comment by DevVersion** [comment]
@fxck What you mean, should be the hover effect.. For now the target of this PR is only making the `ripples` conform with the specs (as specified [here](https://material.io/guidelines/motion/material-motion.html#material-motion-how-does-material-move))

Updating the hover effect of the buttons and other components can be a follow-up.

**Comment by fxck** [comment]
Nonono, it's not the hover effect(it happens after mousedown and I think it's part of the ripple. Want me to make a slowed down video? :)

**Comment by fxck** [comment]
default state
image

hover state(no change)
image

mouse down(notice the grey background)
image

full ripple
image

**Comment by DevVersion** [comment]
That's because Material 1 sets an option to `dim` the background ([see](https://github.com/angular/material/blob/master/src/core/services/ripple/button_ripple.js#L39)). This isn't part of the specification as seen in the [specs](https://material.io/guidelines/motion/material-motion.html#material-motion-how-does-material-move) and also not a big deal either.

> Also add such questions in a PR comment so we can dismiss those. Now the PR is looking unclear...

**Comment by fxck** [comment]
Ok, removed them.

**Comment by DevVersion** [comment]
@fxck Thanks. I've edited my comment to include what you basically commented.

File: src/lib/core/ripple/ripple-renderer.ts [hunk 1]
```diff
@@ -40,167 +29,142 @@ const distanceToFurthestCorner = (x: number, y: number, rect: ClientRect) => {
...
+  /** Whether mouse ripples should be created or not. */
+  rippleDisabled: boolean = false;
+
+  constructor(_elementRef: ElementRef, private _ngZone: NgZone, private _ruler: ViewportRuler) {
+    this._targetElement = _elementRef.nativeElement;
+
+    // Specify events which need to be registered on the trigger.
+    this._triggerEvents.set('mousedown', this.onMousedown.bind(this));
+    this._triggerEvents.set('mouseup', this.onMouseup.bind(this));
+    this._triggerEvents.set('mouseleave', this.onMouseLeave.bind(this));
```
Comment (jelbourn) [inline src/lib/core/ripple/ripple-renderer.ts]: Rather than using a `Map`, how about an array of tuples? Better for iteration
```ts
let triggerEvents: [string, (MouseEvent) => void][] = [
    ['mousedown', e => this.onMouseDown(e)],
    // ...
];
```
(generally prefer arrow functions to `.bind()`)

Comment (DevVersion) [inline src/lib/core/ripple/ripple-renderer.ts]: Don't think that will be better for iteration. A map already provides a suitable iterator for those eventpairs. 

I would personally like to stay with the `Map` for convenience and readability.

**Comment by DevVersion** [comment]
@jelbourn Addressed your feedback. Thanks.

File: src/lib/core/ripple/ripple.spec.ts [hunk 2]
```diff
@@ -86,112 +65,86 @@ describe('MdRipple', () => {
     document.body.style.margin = originalBodyMargin;
   });
 
+  function dispatchMouseEvent(type: string, offsetX = 0, offsetY = 0) {
```
Comment (mmalerba) [inline src/lib/core/ripple/ripple.spec.ts]: I swear this function is in like half of our test files, we should create a test utilities file sometime (not part of this PR though)

File: src/lib/radio/_radio-theme.scss [hunk 3]
```diff
@@ -29,7 +29,7 @@
     }
   }
 
-  .md-radio-ripple .md-ripple-foreground {
+  .md-radio-ripple .md-ripple-element {
```
Comment (mmalerba) [inline src/lib/radio/_radio-theme.scss]: when adding new classes make them mat-* to make andrew's life easier :p

File: src/lib/checkbox/checkbox.html [hunk 4]
```diff
@@ -18,8 +18,7 @@
     <div md-ripple *ngIf="!_isRippleDisabled()" class="md-checkbox-ripple"
          [mdRippleTrigger]="_getHostElement()"
          [mdRippleCentered]="true"
-         [mdRippleSpeedFactor]="0.3"
-         mdRippleBackgroundColor="rgba(0, 0, 0, 0)"></div>
+         [mdRippleSpeedFactor]="0.3"></div>
```
Comment (DevVersion) [inline src/lib/checkbox/checkbox.html]: As discussed on Slack, we are going to have another PR to make all component ripples align with the specs.

File: src/lib/core/ripple/ripple-renderer.ts [hunk 5]
```diff
@@ -40,167 +29,142 @@ const distanceToFurthestCorner = (x: number, y: number, rect: ClientRect) => {
...
-  /** Installs event handlers on the host element of the md-ripple directive. */
-  setTriggerElementToHost() {
-    this.setTriggerElement(this._rippleElement);
-  }
+    let radius = config.radius || distanceToFurthestCorner(pageX, pageY, containerRect);
+    let duration = 1 / (config.speedFactor || 1) * (radius / RIPPLE_SPEED_PX_PER_SECOND);
+    let offsetX = pageX - containerRect.left;
+    let offsetY = pageY - containerRect.top;
+
+    let ripple = document.createElement('div');
```
Comment (DevVersion) [inline src/lib/core/ripple/ripple-renderer.ts]: As discussed on Slack, we will revisit this in issue #308 and once we have feedback from Jeremy about this.

File: src/lib/checkbox/_checkbox-theme.scss [hunk 6]
```diff
@@ -68,16 +68,16 @@
     }
   }
 
-  .md-checkbox:not(.md-checkbox-disabled) {
-    &.md-primary .md-checkbox-ripple .md-ripple-foreground {
+  md-checkbox:not(.md-checkbox-disabled) {
```
Comment (DevVersion) [inline src/lib/checkbox/_checkbox-theme.scss]: Good point.  I actually had that change only in this PR for testing purposes. I have another PR open for this. See https://github.com/angular/material2/pull/2857

Also I think we can only use the class here, as soon as Andrew's PR is merged.

**Comment by DevVersion** [comment]
@mmalerba Thanks for your feedback. Addressed the comments you added. Also glad you like it :)

File: src/lib/checkbox/checkbox.html (hunk 4 above)
Comment (mmalerba) [inline src/lib/checkbox/checkbox.html]: created #2901 for this

Comment (DevVersion) [inline src/lib/checkbox/checkbox.html]: Should I assign it to me?

File: src/lib/core/ripple/ripple.spec.ts (hunk 2 above)
Comment (mmalerba) [inline src/lib/core/ripple/ripple.spec.ts]: created #2902 for this

**Comment by mmalerba** [comment]
lgtm

File: src/lib/core/ripple/ripple-renderer.ts [hunk 7]
```diff
@@ -1,27 +1,9 @@
...
-export class ForegroundRipple {
-  state = ForegroundRippleState.NEW;
-  constructor(public rippleElement: Element) {}
-}
 
-const RIPPLE_SPEED_PX_PER_SECOND = 1000;
-const MIN_RIPPLE_FILL_TIME_SECONDS = 0.1;
-const MAX_RIPPLE_FILL_TIME_SECONDS = 0.3;
+export const RIPPLE_SPEED_PX_PER_SECOND = 170;
+export const RIPPLE_FADE_OUT_DURATION = 600;
```
Comment (tinayuangao) [inline src/lib/core/ripple/ripple-renderer.ts]: Add doc somewhere to mention the fade out duration is fixed and is not changed with speed factor?

Comment (DevVersion) [inline src/lib/core/ripple/ripple-renderer.ts]: Good point. Added it to the `@Input()` and also added comments for those constants.

**Comment by DevVersion** [comment]
Just rebased the PR due to the compatibility changes.

**Comment by realappie** [comment]
Great PR! Was about to open up an issue about the current state of ripple effects.
